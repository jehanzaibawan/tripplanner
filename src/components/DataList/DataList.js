import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import Icon from "@material-ui/core/Icon";
import Divider from "@material-ui/core/Divider";
import ListItemText from "@material-ui/core/ListItemText";
import ListItemAvatar from "@material-ui/core/ListItemAvatar";
import ListItemSecondaryAction from "@material-ui/core/ListItemSecondaryAction";
import Avatar from "@material-ui/core/Avatar";
import Typography from "@material-ui/core/Typography";
import PropTypes from "prop-types";
import {
  durationToMinutes,
  getTimeFromMins,
  getPriceAfterDiscount
} from "../../utils/utils";

const useStyles = makeStyles(theme => ({
  root: {
    width: "100%",
    backgroundColor: theme.palette.background.paper
  },
  inline: {
    display: "inline"
  },
  euro: {
    fontSize: "1rem"
  }
}));

const formatTranportIcons = means => {
  if (means === "car") return <Icon>directions_car</Icon>;
  if (means === "bus") return <Icon>directions_bus</Icon>;
  if (means === "train") return <Icon>train</Icon>;
  else return <Icon>card_travel</Icon>;
};

const getTotalTravellingTimeAndCost = items => {
  return items.reduce(
    (pres, ittr) => {
      return {
        time: (pres.time += durationToMinutes(
          ittr.duration.h,
          ittr.duration.m
        )),
        cost: (pres.cost += getPriceAfterDiscount(ittr.cost, ittr.discount))
      };
    },
    { time: 0, cost: 0 }
  );
};

const formatListHeader = (items, classes) => {
  if (Array.isArray(items) && items.length > 0) {
    const { time, cost } = getTotalTravellingTimeAndCost(items);
    const { hours, minutes } = getTimeFromMins(time);
    return (
      <ListItem button selected={true}>
        <ListItemText
          primary={`${items[0].departure} > ${items[items.length - 1].arrival}`}
          secondary={`Trip duration is ${hours}h ${minutes}m`}
        />
        <ListItemSecondaryAction>
          <Typography
            variant="h6"
            component="h5"
            className={classes.inline}
            color="textPrimary"
          >
            {`Trip cost: ${cost}`}
            <Icon className={classes.euro}>euro_symbol</Icon>
          </Typography>
        </ListItemSecondaryAction>
      </ListItem>
    );
  }
};

const formatListItems = (items, classes) => {
  if (Array.isArray(items) && items.length > 0)
    return items.map((item, index) => (
      <React.Fragment key={`list-item-${index + 1}`}>
        <ListItem alignItems="flex-start">
          <ListItemAvatar>
            <Avatar className={classes.avatar}>
              {formatTranportIcons(item.transport)}
            </Avatar>
          </ListItemAvatar>
          <ListItemText
            primary={`${item.departure} > ${item.arrival}`}
            secondary={
              <React.Fragment>
                <Typography
                  component="span"
                  variant="body2"
                  className={classes.inline}
                  color="textPrimary"
                >
                  {item.transport}
                </Typography>
                {` ${item.reference} for ${item.duration.h}h ${item.duration.m}m`}
              </React.Fragment>
            }
          />
          <ListItemSecondaryAction>
            <Typography
              variant="h6"
              component="h5"
              className={classes.inline}
              color="textPrimary"
            >
              {item.cost}
              <Icon className={classes.euro}>euro_symbol</Icon>
            </Typography>
          </ListItemSecondaryAction>
        </ListItem>
        <Divider variant="inset" component="li" />
      </React.Fragment>
    ));
};

const DataList = props => {
  const classes = useStyles();
  return (
    <List className={classes.root}>
      {formatListHeader(props.data, classes)}
      {formatListItems(props.data, classes)}
    </List>
  );
};

DataList.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({})).isRequired
};

export default DataList;
