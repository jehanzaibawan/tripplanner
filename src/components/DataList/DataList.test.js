import React from "react";
import renderer from "react-test-renderer";
import DataList from "./DataList";

it("renders a snapshot", () => {
  const props = {
    data: [
      {
        transport: "train",
        departure: "London",
        arrival: "Amsterdam",
        duration: {
          h: "05",
          m: "00"
        },
        cost: 160,
        discount: 0,
        reference: "TLA0500"
      }
    ]
  };
  const tree = renderer.create(<DataList {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});
