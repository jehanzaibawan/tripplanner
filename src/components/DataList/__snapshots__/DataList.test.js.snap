// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a snapshot 1`] = `
<ul
  className="MuiList-root makeStyles-root-1 MuiList-padding"
>
  <li
    className="MuiListItem-container"
  >
    <div
      aria-disabled={false}
      className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button MuiListItem-secondaryAction Mui-selected"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiListItemText-root MuiListItemText-multiline"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
        >
          London &gt; Amsterdam
        </span>
        <p
          className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          Trip duration is 5h 0m
        </p>
      </div>
    </div>
    <div
      className="MuiListItemSecondaryAction-root"
    >
      <h5
        className="MuiTypography-root makeStyles-inline-2 MuiTypography-h6 MuiTypography-colorTextPrimary"
      >
        Trip cost: 160
        <span
          aria-hidden={true}
          className="material-icons MuiIcon-root makeStyles-euro-3"
        >
          euro_symbol
        </span>
      </h5>
    </div>
  </li>
  <li
    className="MuiListItem-container"
  >
    <div
      className="MuiListItem-root MuiListItem-gutters MuiListItem-alignItemsFlexStart MuiListItem-secondaryAction"
      disabled={false}
    >
      <div
        className="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart"
      >
        <div
          className="MuiAvatar-root MuiAvatar-colorDefault"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root"
          >
            train
          </span>
        </div>
      </div>
      <div
        className="MuiListItemText-root MuiListItemText-multiline"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
        >
          London &gt; Amsterdam
        </span>
        <p
          className="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          <span
            className="MuiTypography-root makeStyles-inline-2 MuiTypography-body2 MuiTypography-colorTextPrimary"
          >
            train
          </span>
           TLA0500 for 05h 00m
        </p>
      </div>
    </div>
    <div
      className="MuiListItemSecondaryAction-root"
    >
      <h5
        className="MuiTypography-root makeStyles-inline-2 MuiTypography-h6 MuiTypography-colorTextPrimary"
      >
        160
        <span
          aria-hidden={true}
          className="material-icons MuiIcon-root makeStyles-euro-3"
        >
          euro_symbol
        </span>
      </h5>
    </div>
  </li>
  <li
    className="MuiDivider-root MuiDivider-inset"
    role="separator"
  />
</ul>
`;
